### PR TITLE
Check status of the envelope before processing

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1321 (Blob router doesn't fetch envelope from DB after acquiring lease)

### Change description ###
Introduced predicate for checking if status of the envelope being processed is still CREATED and not changed due to any race condition.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessor.java [hunk 1]
```diff
@@ -78,7 +81,11 @@ private void processBlob(BlobClient blob) {
             .ifPresentOrElse(
                 envelope -> {
                     if (envelope.status == Status.CREATED) {
-                        leaseAndThen(blob, () -> blobProcessor.continueProcessing(envelope.id, blob));
+                        leaseAndThen(
+                            blob,
+                            this::isEnvelopeInCreatedStatus,
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessor.java:86]: Why are we involving this check in lease acquiring process?

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessor.java:86]: Also, blobProcessor.process will need to do some custom stuff, too:
- create the envelope and process the blob when there's no envelope
- pick up the existing envelope if it's in the created state
- abandon processing if the envelope is not in created state

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessor.java]: All this is done in LeaseAcquirer. And the check is called not while acquiring the lease, but after that.

File: src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessorTest.java [hunk 2]
```diff
@@ -140,10 +140,10 @@ private void dbHas(Envelope envelope) {
     @SuppressWarnings("unchecked")
     private void leaseCanBeAcquired() {
         doAnswer(invocation -> {
-            var okAction = (Consumer) invocation.getArgument(1);
+            var okAction = (Consumer) invocation.getArgument(2);
             okAction.accept(UUID.randomUUID().toString());
             return null;
-        }).when(leaseAcquirer).ifAcquiredOrElse(any(), any(), any(), anyBoolean());
+        }).when(leaseAcquirer).ifAcquiredOrElse(any(), any(), any(), any(), anyBoolean());
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/ContainerProcessorTest.java:146]: missing new test

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/LeaseAcquirer.java [hunk 3]
```diff
@@ -37,12 +38,30 @@ public void ifAcquiredOrElse(
...
+            blob -> true,
+            onSuccess,
+            onFailure,
+            releaseLease
+        );
+    }
+
+    public void ifAcquiredOrElse(
+        BlobClient blobClient,
+        Predicate<BlobClient> blobCondition,
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/LeaseAcquirer.java:53]: I disagree with this change - it's unnecessary and adds more complexity to the code that should only deal with acquiring leases. Why not just have this condition in the actual business logic?

**Comment by apereverzin** [comment]
https://github.com/hmcts/blob-router-service/pull/589

**Pull request closed without merging**